xtask/src/main.rs

```diff
@@ -0,0 +1,148 @@
+use std::{
+    env, fs,
+    path::{Path, PathBuf},
+    process::Command,
+};
+
+type DynError = Box<dyn std::error::Error>;
+
+fn main() {
+    if let Err(e) = try_main() {
+        eprintln!("{}", e);
+        std::process::exit(-1);
+    }
+}
+
+fn try_main() -> Result<(), DynError> {
+    let task = env::args().nth(1);
+    match task.as_deref() {
+        Some("build") => build()?,
+        _ => print_help(),
+    }
+    Ok(())
+}
+
+fn print_help() {
+    eprintln!(
+        "Tasks:
+
+wasm            build wasm using wasm32-wasi target
+components      build components using wasm-tools
+"
+    )
+}
+
+fn build() -> Result<(), DynError> {
+    let _ = fs::remove_dir_all(&target_dir());
+    fs::create_dir_all(&target_dir())?;
+
+    build_wasm()?;
+    build_components()?;
+
+    Ok(())
+}
+
+fn build_wasm() -> Result<(), DynError> {
+    for guest_name in guest_names() {
+        let cargo = env::var("CARGO").unwrap_or_else(|_| "cargo".to_string());
+        let status = Command::new(cargo)
+            .current_dir(project_root())
+            .args(&[
+                "build",
+                "--target=wasm32-wasi",
+                "--release",
+                &format!("--package={}", guest_name),
+            ])
+            .status()?;
+
+        if !status.success() {
+            Err(format!(
+                "`cargo build --target wasm32-wasi --release --package {}` failed. Status code: {}",
+                guest_name,
+                status.code().unwrap()
+            ))?;
+        }
+    }
+
+    Ok(())
+}
+
+fn build_components() -> Result<(), DynError> {
+    fs::create_dir_all(project_root().join("components"))?;
+
+    for guest_name in guest_names() {
+        build_component(guest_name)?;
+    }
+
+    Ok(())
+}
+
+fn build_component(name: String) -> Result<(), DynError> {
+    let wasm_tools = env::var("WASM_TOOLS").unwrap_or_else(|_| "wasm-tools".to_string());
+
+    let path_wasm_guest = format!(
+        "{}/wasm32-wasi/release/{}.wasm",
+        target_dir().display(),
+        name
+    );
+
+    let path_wasi_snapshot = format!(
```

Comment:
Check this out 

https://doc.rust-lang.org/std/path/struct.Path.html#method.join

I think we'll benefit by using the rust `Path` type. 

Might be able to save ourselves some headache by checking if the path exists. Then, building the command string.